### PR TITLE
enhancement(remap): added is_log and is_metric functions

### PR DESCRIPTION
**Pull request by FungusHumungus**
Ref #4791 

This adds `is_log` and `is_metric` Remap functions so we can deprecate the `is_log` and `is_metric` conditions in favour of Remap.

To work out what type of event we are dealing with I have added an `identifier` function to the `Object` trait that returns a string identifier. The `is_log` function just checks that this string is == "log" and `is_metric` checks == "metric".

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: lib/remap-lang/src/object.rs
Comment (bruceg) [inline lib/remap-lang/src/object.rs:59]: Would it make sense to have this return an enum? I expect this was likely done as a string to allow for extensibility. However, since the `is_log` and `is_metric` functions are part of the remap core, you're already hard-coding the three types into the core.

**Review by bruceg — APPROVED** [review]
Looks as straightforward as expected. One question about the use of strings.

File: lib/remap-lang/src/object.rs
Comment (JeanMertz) [reply lib/remap-lang/src/object.rs:59]: All functions are optional. We load them all in through Vector. VRL itself comes without any functions at the moment.

Comment (FungusHumungus) [reply lib/remap-lang/src/object.rs:59]: > I expect this was likely done as a string to allow for extensibility. 

Precisely. VRL doesn't have to be tied to Vector and `Object` is a core trait of the language. If a third party wanted to use VRL I don't want them to be limited to an Enum we provide. 

I agree, it does feel a bit incongruous having `is_log` and `is_metric` there. 

To make it more generic we could just have a function `object_type()` that returns the identifier. Conditions could then specify `object_type() == "metric"` rather than `is_metric()`. I'm not so keen on that though.

In time perhaps we may split out the functions into Vector specific and generic VRL functions, that would be a non-breaking change we can consider for later.

File: docs/reference/remap/functions/is_log.cue
Comment (lucperkins) [inline docs/reference/remap/functions/is_log.cue:21]: I think it'd be best to provide more realistic, even if contrived, examples for these two functions. Maybe something an `if` statement?

Comment (FungusHumungus) [reply docs/reference/remap/functions/is_log.cue:21]: I'm not quite sure what I could do. These functions are really designed for use mainly with the conditions, so the main source probably would be just `is_log()`, or perhaps `is_log() && contains(.message, "turnips")` - but I'm not sure if that makes for a good example..

**Review by JeanMertz — COMMENTED** [review]
I’m not convinced this is the right approach to take, but I also haven’t had the opportunity to think through the implications and alternatives.

I wanted to voice this concern before it is merged, but if we deem these functions necessary before 0.12 is released, then I don’t want to hold this PR up.

File: docs/reference/remap/functions/is_log.cue
Comment (binarylogic) [reply docs/reference/remap/functions/is_log.cue:21]: Yes, we should support an `input` with a log event for this example, since it operates directly on the event. Ex:

```
input: log: message: "Hello, world!"
```

File: docs/reference/remap/functions/is_metric.cue
Comment (binarylogic) [inline docs/reference/remap/functions/is_metric.cue:23]: Same here, we should add a metric event. Ex:

```
input: metric: {
	kind: "incremental"
	name: "user_login_total"
	counter: value: 102.0
	tags: host: "my.host.com"
}
```

File: docs/reference/remap/functions/is_metric.cue
Comment (binarylogic) [inline docs/reference/remap/functions/is_metric.cue:13]:
```suggestion
	category: "Event"
```

File: docs/reference/remap/functions/is_log.cue
Comment (binarylogic) [inline docs/reference/remap/functions/is_log.cue:13]:
```suggestion
	category: "Event"
```

File: docs/reference/remap/functions/is_log.cue
Comment (binarylogic) [inline docs/reference/remap/functions/is_log.cue:10]:
```suggestion
			"If the current event is a [`log` event](\(urls.vector_log)), then `true` is returned.",
```

File: docs/reference/remap/functions/is_log.cue
Comment (binarylogic) [inline docs/reference/remap/functions/is_log.cue:16]:
```suggestion
	description: #"""
		Determines whether the current event is a [`log` event](\(urls.vector_log)).
		"""#
```

File: docs/reference/remap/functions/is_metric.cue
Comment (binarylogic) [inline docs/reference/remap/functions/is_metric.cue:10]:
```suggestion
			"If the current event is a [`metric` event](\(urls.vector_metric)), then `true` is returned."#,
```

File: docs/reference/remap/functions/is_metric.cue
Comment (binarylogic) [inline docs/reference/remap/functions/is_metric.cue:16]:
```suggestion
	description: """
		Determines whether the current event is a [`metric` event](\(urls.vector_metric)).
		"""
```

**Comment by FungusHumungus** [comment]
> I’m not convinced this is the right approach to take, but I also haven’t had the opportunity to think through the implications and alternatives.

Is it the API (having the two functions `is_log` and `is_metric`) that you aren't keen on? Or is it the way they are implemented?

**Comment by FungusHumungus** [comment]
Talking with @JeanMertz, we have decided to put this off for now and try to think of a better way that this can be handled. Conditions will still work with `condition.type = "is_metric"`.

**Pull request closed without merging**
